program/src/processor.rs

```diff
@@ -5673,6 +5668,150 @@ impl Processor {
+        let _ = MangoGroup::load_checked(mango_group_ai, program_id)?;
+        let mango_account =
+            MangoAccount::load_checked(mango_account_ai, program_id, &mango_group_ai.key)?;
+        check!(
+            &mango_account.owner == owner_ai.key || &mango_account.delegate == owner_ai.key,
+            MangoErrorCode::InvalidOwner
+        )?;
+        check!(owner_ai.is_signer, MangoErrorCode::InvalidSignerKey)?;
+
+        let _ =
+            MangoAccount::load_checked(referrer_mango_account_ai, program_id, mango_group_ai.key)?;
+
+        if referrer_memory_ai.data_is_empty() {
+            // initialize it if it's not initialized yet
+            let referrer_seeds: &[&[u8]] = &[&mango_account_ai.key.as_ref(), b"ReferrerMemory"];
+            seed_and_create_pda(
+                program_id,
+                payer_ai,
+                &Rent::get()?,
+                size_of::<ReferrerMemory>(),
+                program_id,
+                system_prog_ai,
+                referrer_memory_ai,
+                referrer_seeds,
+                &[],
+            )?;
+        }
+
+        let mut referrer_memory = ReferrerMemory::load_mut_checked(referrer_memory_ai, program_id)?;
+        referrer_memory.referrer_mango_account = *referrer_mango_account_ai.key;
+
+        Ok(())
+    }
+
+    /// Associate the referrer's MangoAccount with a human readable `referrer_id` which can be used
+    /// in a ref link
+    /// Create the `ReferrerIdRecord` PDA; if it already exists throw error
+    #[inline(never)]
+    fn register_referrer_id(
+        program_id: &Pubkey,
+        accounts: &[AccountInfo],
+        referrer_id: [u8; INFO_LEN],
+    ) -> MangoResult {
+        const NUM_FIXED: usize = 6;
+        let [
+            mango_group_ai,             // read
+            referrer_mango_account_ai,  // read
+            owner_ai,                   // signer
+            referrer_id_record_ai,      // write
+            payer_ai,                   // write, signer
+            system_prog_ai,             // read
+        ] = array_ref![accounts, 0, NUM_FIXED];
+        check!(
+            system_prog_ai.key == &solana_program::system_program::id(),
+            MangoErrorCode::InvalidProgramId
+        )?;
+
+        let _ = MangoGroup::load_checked(mango_group_ai, program_id)?;
+
+        let referrer_mango_account =
+            MangoAccount::load_checked(referrer_mango_account_ai, program_id, mango_group_ai.key)?;
+        check!(
+            &referrer_mango_account.owner == owner_ai.key
+                || &referrer_mango_account.delegate == owner_ai.key,
+            MangoErrorCode::InvalidOwner
+        )?;
+        check!(owner_ai.is_signer, MangoErrorCode::InvalidSignerKey)?;
+
+        // referrer_id_record must be empty; cannot be transferred
+        check!(referrer_id_record_ai.data_is_empty(), MangoErrorCode::InvalidAccount)?;
+        let referrer_record_seeds: &[&[u8]] = &[&mango_group_ai.key.as_ref(), &referrer_id];
```

Comment:
important: Add a fixed identifier between mango_group_ai and referrer_id. For example the seed `[mango_group_ai, "DustAccount"]` is already taken.